chasm/lib/activity/handler.go

```diff
@@ -51,3 +54,115 @@ func (h *handler) StartActivityExecution(ctx context.Context, req *activitypb.St
 		FrontendResponse: response,
 	}, nil
 }
+
+// PollActivityExecution handles PollActivityExecutionRequest from frontend. This method supports
+// querying current activity state, optionally as a long-poll that waits for certain state changes.
+// It is used by clients to poll for activity state and/or result.
+func (h *handler) PollActivityExecution(
+	ctx context.Context,
+	req *activitypb.PollActivityExecutionRequest,
+) (response *activitypb.PollActivityExecutionResponse, err error) {
+	ref := chasm.NewComponentRef[*Activity](chasm.EntityKey{
+		NamespaceID: req.GetNamespaceId(),
+		BusinessID:  req.GetFrontendRequest().GetActivityId(),
+		EntityID:    req.GetFrontendRequest().GetRunId(),
+	})
+	defer func() {
+		var notFound *serviceerror.NotFound
+		if errors.As(err, &notFound) {
+			// TODO(dan): include execution key in error message; we may do this at the CHASM
+			// framework level.
+			// cf. "workflow execution not found for workflow ID XXX and run ID YYY"
+			err = serviceerror.NewNotFound("activity execution not found")
+		}
+	}()
+
+	waitPolicy := req.GetFrontendRequest().GetWaitPolicy()
+
+	if waitPolicy == nil {
+		return chasm.ReadComponent(ctx, ref, (*Activity).buildPollActivityExecutionResponse, req, nil)
+	}
+
+	// Do not allow long-poll to use all remaining time
+	childCtx := ctx
+	if deadline, ok := ctx.Deadline(); ok {
```

Comment:
Use the helper from `contextutil` and cap the wait timeout to a minute like other long poll APIs.